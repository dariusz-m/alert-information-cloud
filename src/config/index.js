import devConfig from "./config.development";
import prodConfig from "./config.production";
import testConfig from "./config.test";

const env = process.env.NODE_ENV || "development";

const config = {
    "development": devConfig,
    "production": prodConfig,
    "test": testConfig,
};

export default config[env];
