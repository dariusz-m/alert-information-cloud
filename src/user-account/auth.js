export const isUserAuthenticated = () => {
    return sessionStorage.getItem("sessionId") !== null;
};
