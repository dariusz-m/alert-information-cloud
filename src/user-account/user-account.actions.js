export const AccountActionType = {
    LOGIN: 'LOGIN',
    LOGOUT: 'LOGOUT',
    USER_WENT_TO_LOGIN_PAGE: 'USER_WENT_TO_LOGIN_PAGE',
    USER_HAS_BEEN_LOGGED_IN: 'USER_HAS_BEEN_LOGGED_IN',
    USER_HAS_NOT_BEEN_LOGGED_IN: 'USER_HAS_NOT_BEEN_LOGGED_IN',

    REGISTER: 'REGISTER',
    USER_WENT_TO_REGISTER_PAGE: 'USER_WENT_TO_REGISTER_PAGE',
    USER_HAS_BEEN_REGISTERED: 'USER_HAS_BEEN_REGISTERED',
    USER_HAS_NOT_BEEN_REGISTERED: 'USER_HAS_NOT_BEEN_REGISTERED',
};

export const login = (email, password) => {
    return {
        type: AccountActionType.LOGIN,
        payload: {
            email,
            password,
        }
    };
};

export const logout = () => {
    return {
        type: AccountActionType.LOGOUT,
    };
};

export const userWentToLoginPage = () => {
    return {
        type: AccountActionType.USER_WENT_TO_LOGIN_PAGE,
    };
};


export const userHasBeenLoggedIn = () => {
    return {
        type: AccountActionType.USER_HAS_BEEN_LOGGED_IN
    };
};

export const userHasNotBeenLoggedIn = (errorMessage) => {
    return {
        type: AccountActionType.USER_HAS_NOT_BEEN_LOGGED_IN,
        payload: errorMessage
    };
};

export const register = (email, password, name, surname) => {
    return {
        type: AccountActionType.REGISTER,
        payload: {
            email,
            password,
            name,
            surname
        }
    };
};

export const userWentToRegisterPage = () => {
    return {
        type: AccountActionType.USER_WENT_TO_REGISTER_PAGE,
    };
};

export const userHasBeenRegistered = () => {
    return {
        type: AccountActionType.USER_HAS_BEEN_REGISTERED
    };
};

export const userHasNotBeenRegistered = (errorMessage) => {
    return {
        type: AccountActionType.USER_HAS_NOT_BEEN_REGISTERED,
        payload: errorMessage
    };
};

