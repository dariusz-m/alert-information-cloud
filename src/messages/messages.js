import "./messages.scss";

import React from "react";
import {connect} from "react-redux";
import PropTypes from "prop-types";
import {bindActionCreators} from "redux";
import ReactList from "react-list";
import TimeAgo from "react-timeago";

import MultiSelect from "../multiselect/multiSelect";
import {sendMessage, subscribeMessagesChannel} from "./messages.actions";
import {MessageNotification} from "../notification/notification";
import {logout} from "../user-account/user-account.actions";

/**
 * Main component, handling messages
 */
export class Messages extends React.Component {
    /**
     * Constructor.
     *
     * @param {Object} props: Component configuration.
     */
    constructor(props) {
        super(props);

        this.state = {
            selectedUsersIds: [],
            message: "",
            selectedMessage: null
        };

        this.onSelectedOptionsChange = this.onSelectedOptionsChange.bind(this);
        this.onMessageBodyChange = this.onMessageBodyChange.bind(this);
        this.sendMessage = this.sendMessage.bind(this);
        this.renderMessage = this.renderMessage.bind(this);
        this.displayMessageDetails = this.displayMessageDetails.bind(this);
    }

    /**
     * On selected options change
     *
     * @param {Array} selectedUsersIds: Selected users ids
     */
    onSelectedOptionsChange(selectedUsersIds) {
        this.setState({
            selectedUsersIds: selectedUsersIds
        });
    }

    /**
     * On message body change.
     *
     * @param {Event} changeEvent: Change event.
     */
    onMessageBodyChange(changeEvent) {
        this.setState({message: changeEvent.target.value});
    }

    /**
     * On message body change.
     *
     * @param {Event} clickEvent: Click event.
     */
    sendMessage(clickEvent) {
        if (this.state.selectedUsersIds.length > 0) {
            this.props.actions.sendMessage(this.state.message, this.state.selectedUsersIds);
        }
        clickEvent.preventDefault();
    }

    /**
     * Component did mount.
     */
    componentDidMount() {
        this.props.actions.subscribeMessagesChannel();
    }

    /**
     * Display message details
     *
     * @param {Object} message: Message
     * */
    displayMessageDetails(message) {
        this.setState({selectedMessage: message});
    }

    /**
     * Render component.
     *
     * @param {number} index: Index of message
     *
     * @returns {React.Element}
     */
    renderMessage(index) {
        return (
            <div
                key={this.props.messages[index].timestamp}
                className="result-block"
                onClick={this.displayMessageDetails.bind(this, this.props.messages[index])}
            >
                <div className={`data-piece`}>
                    <span className="result-title-side">
                        <TimeAgo date={this.props.messages[index].timestamp}/>
                    </span>
                </div>
                <div className={`data-piece`}>
                    <span className={`icon flow from`} id="icon-direction"/>
                    <a className="link result-attr-2"> {this.props.messages[index].from}</a>
                    <span className="icon href"/>
                </div>
                <div className={`data-piece`}>
                    <a className="link result-attr-3 truncate">
                        <span className="icon hashtag"/>
                        {this.props.messages[index].message}</a>
                    <span className="icon href"/>
                </div>
            </div>

        );
    }

    /**
     * Render component.
     *
     * @returns {React.Element}
     */
    render() {
        const options = this.props.users.map((user) => {
            return {id: user.id, text: user.email};
        });

        return (
            <div>
                <button className={"btn btn-danger float-right"} onClick={this.props.actions.logout}>Logout</button>
                <div className={"md-12 messages"}>
                    <div className="scrollable-wrapper">
                        <ReactList
                            itemRenderer={this.renderMessage}
                            length={this.props.messages.length}
                            pageSize={1}
                        />
                    </div>
                    {this.state.selectedMessage ?
                        <div className={"selected-message"}>
                            <h4><strong>{this.state.selectedMessage.from}</strong> -- sent you message</h4>
                            <p>{this.state.selectedMessage.message}</p>
                            <span>{new Date(this.state.selectedMessage.timestamp).toString()}</span>
                        </div>
                        : null
                    }
                </div>
                <div className={"md-12"}>
                    <MultiSelect
                        options={options}
                        selectedOptionsIds={this.state.selectedUsersIds}
                        updateSelectedOptions={this.onSelectedOptionsChange}
                    />
                    <div className="input-group mb-3">
                        <input
                            type="text" className="form-control" placeholder="Message body..."
                            aria-label="Message body..." aria-describedby="basic-addon2"
                            value={this.state.message} onChange={this.onMessageBodyChange}
                        />
                        <div className="input-group-append">
                            <button
                                className="btn btn-outline-secondary"
                                type="button"
                                disabled={!this.state.message}
                                onClick={this.sendMessage}
                            >
                                Send
                            </button>
                        </div>
                    </div>
                    <div className="md-12">
                        {(() => {
                            switch(this.props.messageHasBeenSent) {
                                case false:
                                    return(
                                        <div className="alert alert-danger">
                                            <div>
                                                <strong>Something went wrong!</strong> Try again.
                                            </div>
                                        </div>
                                    );
                                case true:
                                    return(
                                        <div className="alert alert-success">
                                            <strong>Success!</strong> Message has been sent.
                                        </div>
                                    );
                                default:
                                    return this.state.selectedUsersIds.length === 0 ?
                                        <div className="alert alert-info">
                                            <p>Select at least one recipient to send message.</p>
                                        </div>
                                        : null;

                            }
                        })()}
                    </div>
                </div>
                <MessageNotification message={this.props.lastMessage}/>
            </div>
        );
    }
}


Messages.propTypes = {
    users: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.number.isRequired,
        email: PropTypes.string.isRequired,
        firstname: PropTypes.string.isRequired,
        lastname: PropTypes.string.isRequired,
    })).isRequired,
    messages: PropTypes.arrayOf(PropTypes.shape({
        timestamp: PropTypes.number.isRequired,
        message: PropTypes.string.isRequired,
        from: PropTypes.string.isRequired,
    })).isRequired,
    lastMessage: PropTypes.shape({
        timestamp: PropTypes.number.isRequired,
        message: PropTypes.string.isRequired,
        from: PropTypes.string.isRequired,
    }),
    messageHasBeenSent: PropTypes.bool,
    actions: PropTypes.shape({
        sendMessage: PropTypes.func.isRequired,
        subscribeMessagesChannel: PropTypes.func.isRequired
    }),
};

/**
 * Map state to props.
 *
 * @param {Object} state: State of application
 *
 * @returns {Object}
 */
function mapStateToProps(state) {
    return {
        users: state.users,
        messages: state.alertInformationPage.messages,
        lastMessage: state.alertInformationPage.lastMessage,
        messageHasBeenSent: state.alertInformationPage.messageHasBeenSent,
    };
}

/**
 * Map state to props.
 *
 * @param {Object} dispatch: Dispatch
 *
 * @returns {Object}
 */
function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            sendMessage: sendMessage,
            subscribeMessagesChannel: subscribeMessagesChannel,
            logout: logout,
        }, dispatch)
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Messages);
