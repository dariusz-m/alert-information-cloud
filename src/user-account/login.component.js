import './user-account.scss';

import React from 'react';
import PropTypes from 'prop-types';
import {connect} from "react-redux";
import {bindActionCreators} from "redux";
import {Link} from "react-router-dom";

import {login} from "./user-account.actions";

/**
 * Login component.
 */
export class Login extends React.Component {
    /**
     * Constructor.
     *
     * @param {Object} props: Component configuration.
     */
    constructor(props) {
        super(props);

        this.state = {
            email: "",
            password: "",
        };

        this.emailChange = this.emailChange.bind(this);
        this.passwordChange = this.passwordChange.bind(this);
        this.login = this.login.bind(this);
    }

    /**
     * On email change.
     *
     * @param {Event} changeEvent: Change event.
     */
    emailChange(changeEvent) {
        this.setState({email: changeEvent.target.value});
    }
    /**
     * On password change.
     *
     * @param {Event} changeEvent: Change event.
     */
    passwordChange(changeEvent) {
        this.setState({password: changeEvent.target.value});
    }
    /**
     * Login.
     *
     * @param {Event} clickEvent: Click event.
     */
    login(clickEvent) {
        this.props.actions.login(this.state.email, this.state.password);
        clickEvent.preventDefault();
    }

    /**
     * Render component.
     *
     * @returns {React.Element}
     */
    render() {
        return (
            <form className="form-signin">
                <img className="mb-4" src="" alt="" width="72" height="72"/>
                <h1 className="h3 mb-3 font-weight-normal">Please sign in</h1>
                <label htmlFor="inputEmail" className="sr-only">Email address</label>
                <input type="email" id="inputEmail" className="form-control" placeholder="Email address" required
                    autoFocus value={this.state.email} onChange={this.emailChange}/>
                <label htmlFor="inputPassword" className="sr-only">Password</label>
                <input type="password" id="inputPassword" className="form-control" placeholder="Password"
                    required value={this.state.password} onChange={this.passwordChange}/>
                <p>No account? <Link to={"/register"}><strong>Create one!</strong></Link></p>
                <button className="btn btn-lg btn-primary btn-block" type="submit" onClick={this.login}>Sign in</button>
                {this.props.errorMessage?
                    <div className="alert alert-danger">
                        <strong>Login failed!</strong> {this.props.errorMessage}
                    </div> : null
                }
            </form>
        );
    }
}

Login.propTypes = {
    errorMessage: PropTypes.string,
    actions: PropTypes.shape({
        login: PropTypes.func.isRequired,
    }),
};

/**
 * Map state to props.
 *
 * @param {Object} state: State of application
 *
 * @returns {Object}
 */
function mapStateToProps(state) {
    return {
        errorMessage: state.userAccount.loginPage.errorMessage,
    };
}

/**
 * Map state to props.
 *
 * @param {Object} dispatch: Dispatch
 *
 * @returns {Object}
 */
function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            login: login,
        }, dispatch)
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Login);
