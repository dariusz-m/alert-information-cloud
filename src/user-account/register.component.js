import './user-account.scss';

import React from 'react';
import PropTypes from 'prop-types';
import {connect} from "react-redux";
import {bindActionCreators} from "redux";
import {Link} from "react-router-dom";

import {register} from "./user-account.actions";

/**
 * Register component.
 */
export class Register extends React.Component {
    /**
     * Constructor.
     *
     * @param {Object} props: Component configuration.
     */
    constructor(props) {
        super(props);

        this.state = {
            email: "",
            name: "",
            surname: "",
            password: ""
        };

        this.emailChange = this.emailChange.bind(this);
        this.nameChange = this.nameChange.bind(this);
        this.surnameChange = this.surnameChange.bind(this);
        this.passwordChange = this.passwordChange.bind(this);
        this.register = this.register.bind(this);
    }

    /**
     * Component will receive props
     *
     * @param {Object} nextProps: Next component configuration.
     */
    componentWillReceiveProps(nextProps) {
        if(nextProps.isUserRegistered) {
            this.setState({
                email: "",
                name: "",
                surname: "",
                password: ""
            });
        }
    }

    /**
     * On email change.
     *
     * @param {Event} changeEvent: Change event.
     */
    emailChange(changeEvent) {
        this.setState({email: changeEvent.target.value});
    }
    /**
     * On name change.
     *
     * @param {Event} changeEvent: Change event.
     */
    nameChange(changeEvent) {
        this.setState({name: changeEvent.target.value});
    }
    /**
     * On name change.
     *
     * @param {Event} changeEvent: Change event.
     */
    surnameChange(changeEvent) {
        this.setState({surname: changeEvent.target.value});
    }
    /**
     * On password change.
     *
     * @param {Event} changeEvent: Change event.
     */
    passwordChange(changeEvent) {
        this.setState({password: changeEvent.target.value});
    }
    /**
     * Register.
     *
     * @param {Event} clickEvent: Click event.
     */
    register(clickEvent) {
        this.props.actions.register(this.state.email, this.state.password, this.state.name, this.state.surname);
        clickEvent.preventDefault();
    }

    /**
     * Render component.
     *
     * @returns {React.Element}
     */
    render() {
        return (
            <form className="form-signin">
                <img className="mb-4" src="" alt="" width="72" height="72"/>
                <h1 className="h3 mb-3 font-weight-normal">Create account</h1>
                <label htmlFor="inputEmail" className="sr-only">Email address</label>
                <input type="email" id="inputEmail" className="form-control" placeholder="Email address" required
                    autoFocus value={this.state.email} onChange={this.emailChange}/>
                <label htmlFor="inputPassword" className="sr-only">Password</label>
                <input type="password" id="inputPassword" className="form-control" placeholder="Password"
                    required value={this.state.password} onChange={this.passwordChange}/>
                <label htmlFor="inputName" className="sr-only">Name</label>
                <input type="text" id="inputName" className="form-control" placeholder="Name"
                    required value={this.state.name} onChange={this.nameChange}/>
                <label htmlFor="inputSurname" className="sr-only">Surname</label>
                <input type="text" id="inputSurname" className="form-control" placeholder="Surname"
                    required value={this.state.surname} onChange={this.surnameChange}/>
                <p>Back to <Link to={"/login"}><strong>Login page</strong></Link></p>
                <button className="btn btn-lg btn-primary btn-block" type="submit" onClick={this.register}>
                    Sign up
                </button>
                {this.props.errorMessage?
                    <div className="alert alert-danger">
                        <strong>Registration failed!</strong> {this.props.errorMessage}
                    </div> : null
                }
                {this.props.isUserRegistered?
                    <div className="alert alert-success">
                        <strong>Success!</strong> You have been registered.
                    </div>: null
                }
            </form>
        );
    }
}

Register.propTypes = {
    errorMessage: PropTypes.string,
    isUserRegistered: PropTypes.bool,
    actions: PropTypes.shape({
        register: PropTypes.func.isRequired,
    }),
};

/**
 * Map state to props.
 *
 * @param {Object} state: State of application
 *
 * @returns {Object}
 */
function mapStateToProps(state) {
    return {
        errorMessage: state.userAccount.registerPage.errorMessage,
        isUserRegistered: state.userAccount.registerPage.isUserRegistered,
    };
}

/**
 * Map state to props.
 *
 * @param {Object} dispatch: Dispatch
 *
 * @returns {Object}
 */
function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            register: register,
        }, dispatch)
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Register);
