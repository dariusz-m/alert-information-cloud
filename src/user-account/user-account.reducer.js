import initialState from '../state/initialState';
import {AccountActionType} from './user-account.actions';

export default (state = initialState.userAccount, action) => {
    switch (action.type) {
        case AccountActionType.USER_HAS_NOT_BEEN_LOGGED_IN:
            return {...state, loginPage: {...state.loginPage, errorMessage: action.payload}};
        case AccountActionType.USER_WENT_TO_LOGIN_PAGE:
            return {...state, loginPage: {...state.loginPage, errorMessage: ""}};

        case AccountActionType.USER_HAS_BEEN_REGISTERED:
            return {...state, registerPage: {...state.registerPage, errorMessage: "", isUserRegistered: true}};
        case AccountActionType.USER_HAS_NOT_BEEN_REGISTERED:
            return {
                ...state,
                registerPage: {...state.registerPage, errorMessage: action.payload, isUserRegistered: false}
            };
        case AccountActionType.USER_WENT_TO_REGISTER_PAGE:
            return {...state, registerPage: {...state.registerPage, errorMessage: "", isUserRegistered: false}};


        default:
            return state;
    }
};
