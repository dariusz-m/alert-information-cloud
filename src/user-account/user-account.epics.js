import { from, of } from 'rxjs';
import { mergeMap, map, mapTo, catchError, pluck, filter, tap} from 'rxjs/operators';
import { ajax } from 'rxjs/ajax';

import config from "../config";
import {
    AccountActionType,
    userHasNotBeenLoggedIn,
    userHasBeenRegistered,
    userHasNotBeenRegistered, userWentToLoginPage, userWentToRegisterPage
} from './user-account.actions';
import {generateKeys} from "../security/generateKeys";
import {LOCATION_CHANGE, push} from "react-router-redux";
import {clearAllTheMessages} from "../messages/messages.actions";

export const login = action$ =>
    action$.ofType(AccountActionType.LOGIN)
        .pipe(
            map((action) => {
                const keys = generateKeys();
                sessionStorage.setItem("clientPrivateKey", keys.privateKey);
                return {
                    ...action,
                    payload: {
                        ...action.payload,
                        clientPublicKey: keys.publicKey,
                    }
                };
            }),
            mergeMap((action) =>
                from(
                    ajax({
                        url: `${config.API_ORIGIN}/api/v1/login`,
                        method: "POST",
                        headers: {'Content-Type': 'application/json'},
                        crossDomain: true,
                        withCredentials: false,
                        body: {
                            email: action.payload.email,
                            password: action.payload.password,
                            client_public_key: action.payload.clientPublicKey, // eslint-disable-line camelcase
                        }
                    })
                ).pipe(
                    pluck('response'),
                    map((data) => {
                        sessionStorage.setItem("serverPublicKey", data.server_public_key);
                        sessionStorage.setItem("sessionId", data.session_id);
                        return push("/");
                    }),
                    catchError((error) => {
                        sessionStorage.clear();
                        return of(userHasNotBeenLoggedIn(error.xhr.response.message));
                    })
                )
            ),
        );


export const logout = action$ =>
    action$.ofType(AccountActionType.LOGOUT)
        .pipe(
            mergeMap(() =>
                from(
                    ajax({
                        url: `${config.API_ORIGIN}/api/v1/user/logout`,
                        method: "GET",
                        headers: {
                            'Content-Type': 'application/json',
                            'Authorization': sessionStorage.getItem("sessionId")
                        },
                        crossDomain: true,
                        withCredentials: false,
                    })
                ).pipe(
                    mergeMap(() => {
                        sessionStorage.clear();
                        return [push("/login"), clearAllTheMessages()];
                    }),
                    catchError(() => {
                        return of(userHasNotBeenLoggedIn(""));
                    })
                )
            ),
        );

export const userWentToLoginPageEpic = action$ => {
    return action$.ofType(LOCATION_CHANGE)
        .pipe(
            filter((action) => action.payload.pathname === "/login"),
            tap(() => sessionStorage.clear()),
            mapTo(userWentToLoginPage())
        );
};

export const userWentToRegisterPageEpic = action$ => {
    return action$.ofType(LOCATION_CHANGE)
        .pipe(
            filter((action) => action.payload.pathname === "/register"),
            tap(() => sessionStorage.clear()),
            mapTo(userWentToRegisterPage())
        );
};

export const register = action$ =>
    action$.ofType(AccountActionType.REGISTER)
        .pipe(
            mergeMap((action) =>
                from(
                    ajax({
                        url: `${config.API_ORIGIN}/api/v1/register`,
                        method: "POST",
                        headers: {'Content-Type': 'application/json'},
                        crossDomain: true,
                        withCredentials: false,
                        body: {
                            email: action.payload.email,
                            firstname: action.payload.name,
                            lastname: action.payload.surname,
                            password: action.payload.password
                        }
                    })
                ).pipe(
                    mapTo(userHasBeenRegistered()),
                    catchError((error) => {
                        return of(userHasNotBeenRegistered(error.xhr.response.message));
                    })
                )
            ),
        );
