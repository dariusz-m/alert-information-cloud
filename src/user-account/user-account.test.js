import React from 'react';
import renderer from "react-test-renderer";
import nock from 'nock';
import {Router} from "react-router-dom";
import createHistory from 'history/createBrowserHistory';
import { ActionsObservable } from 'redux-observable';
import {toArray} from 'rxjs/operators';

import config from "../config";
import configureStore from "../state";
import initialState from "../state/initialState";
import {login, logout, register} from "./user-account.actions";
import {login as loginEpic, logout as logoutEpic} from "./user-account.epics";
import {Login} from "./login.component";
import {Register} from "./register.component";
import {push} from "react-router-redux";
import {clearAllTheMessages} from "../messages/messages.actions";

const keys = {
    publicKey: "340d2a00048539c2329e276488cdca842b78c60a50f3f3721cd180c6f9486d38",
    privateKey: "043282c0159db1e80b95ca901dd876ce7819b80a86fca87df50fb890f3bac3d038b" +
        "6bee3dbbfd96d7e3be6bf7a65a9e9a21b6dddcfaa5cb27f146de6084b3a2df3"
};
/**
 * Mock generate keys.
 *
 * @returns {Object}
 */
function mockGenerateKeys() {
    return {
        generateKeys: jest.fn().mockImplementation(() => keys),
    };
}
jest.mock('../security/generateKeys', () => mockGenerateKeys());

describe('User Account tests', () => {
    let store;
    const history = createHistory();

    beforeEach(() => {
        store = configureStore();
    });

    it('Login component renders correctly', () => {
        const actions = {
            login: () => {},
        };
        const tree = renderer.create(
            <Router history={history}>
                <Login actions={actions}/>
            </Router>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Register component renders correctly', () => {
        const actions = {
            register: () => {},
        };
        const tree = renderer.create(
            <Router history={history}>
                <Register actions={actions}/>
            </Router>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('User can logout', (done) => {
        expect.hasAssertions();

        sessionStorage.setItem("sessionId", "sessionId");
        sessionStorage.setItem("clientPrivateKey", "private key");
        sessionStorage.setItem("serverPublicKey", "public key");
        nock(config.API_ORIGIN)
            .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
            .get(`/api/v1/user/logout`)
            .reply(200, {});

        const expectedOutputActions = [push("/login"), clearAllTheMessages()];
        const action$ = ActionsObservable.of(logout());

        logoutEpic(action$).pipe(
            toArray()
        ).subscribe(actualOutputActions => {
            expect(actualOutputActions).toEqual(expectedOutputActions);
            expect(sessionStorage.getItem("clientPrivateKey")).toEqual(null);
            expect(sessionStorage.getItem("serverPublicKey")).toEqual(null);
            expect(sessionStorage.getItem("sessionId")).toEqual(null);
            done();
        });
    });


    it('User can login in', (done) => {
        expect.hasAssertions();

        const expectedSessionId = "sessionId";
        const response = {
            "server_public_key": keys.publicKey,
            "session_id": expectedSessionId,
            "user": {
                "email": "string",
                "firstname": "string",
                "id": 0,
                "lastname": "string"
            }
        };
        nock(config.API_ORIGIN)
            .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
            .post(`/api/v1/login`)
            .reply(200, response);
        const expectedOutputActions = [
            push("/"),
        ];
        const action$ = ActionsObservable.of(
            login("mail@gmail.com", "password")
        );

        loginEpic(action$).pipe(
            toArray()
        ).subscribe(actualOutputActions => {
            expect(actualOutputActions).toEqual(expectedOutputActions);
            expect(sessionStorage.getItem("clientPrivateKey")).toEqual(keys.privateKey);
            expect(sessionStorage.getItem("serverPublicKey")).toEqual(keys.publicKey);
            expect(sessionStorage.getItem("sessionId")).toEqual(expectedSessionId);
            done();
        });
    });

    it('User can register in our system', (done) => {
        expect.hasAssertions();
        nock(config.API_ORIGIN)
            .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
            .post(`/api/v1/register`)
            .reply(200, {data: {}});
        const expectedState = {
            ...initialState,
            userAccount: {
                ...initialState.userAccount,
                registerPage: {
                    ...initialState.userAccount.registerPage,
                    isUserRegistered: true
                }
            }
        };

        store.dispatch(register("mail@gmail.com", "password", "name", "surname"));

        store.subscribe(() => {
            expect(store.getState()).toEqual(expectedState);
            done();
        });
    });

});

