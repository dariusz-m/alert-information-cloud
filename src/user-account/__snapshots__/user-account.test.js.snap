// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Account tests Login component renders correctly 1`] = `
<form
  className="form-signin"
>
  <img
    alt=""
    className="mb-4"
    height="72"
    src=""
    width="72"
  />
  <h1
    className="h3 mb-3 font-weight-normal"
  >
    Please sign in
  </h1>
  <label
    className="sr-only"
    htmlFor="inputEmail"
  >
    Email address
  </label>
  <input
    autoFocus={true}
    className="form-control"
    id="inputEmail"
    onChange={[Function]}
    placeholder="Email address"
    required={true}
    type="email"
    value=""
  />
  <label
    className="sr-only"
    htmlFor="inputPassword"
  >
    Password
  </label>
  <input
    className="form-control"
    id="inputPassword"
    onChange={[Function]}
    placeholder="Password"
    required={true}
    type="password"
    value=""
  />
  <p>
    No account? 
    <a
      href="/register"
      onClick={[Function]}
    >
      <strong>
        Create one!
      </strong>
    </a>
  </p>
  <button
    className="btn btn-lg btn-primary btn-block"
    onClick={[Function]}
    type="submit"
  >
    Sign in
  </button>
</form>
`;

exports[`User Account tests Register component renders correctly 1`] = `
<form
  className="form-signin"
>
  <img
    alt=""
    className="mb-4"
    height="72"
    src=""
    width="72"
  />
  <h1
    className="h3 mb-3 font-weight-normal"
  >
    Create account
  </h1>
  <label
    className="sr-only"
    htmlFor="inputEmail"
  >
    Email address
  </label>
  <input
    autoFocus={true}
    className="form-control"
    id="inputEmail"
    onChange={[Function]}
    placeholder="Email address"
    required={true}
    type="email"
    value=""
  />
  <label
    className="sr-only"
    htmlFor="inputPassword"
  >
    Password
  </label>
  <input
    className="form-control"
    id="inputPassword"
    onChange={[Function]}
    placeholder="Password"
    required={true}
    type="password"
    value=""
  />
  <label
    className="sr-only"
    htmlFor="inputName"
  >
    Name
  </label>
  <input
    className="form-control"
    id="inputName"
    onChange={[Function]}
    placeholder="Name"
    required={true}
    type="text"
    value=""
  />
  <label
    className="sr-only"
    htmlFor="inputSurname"
  >
    Surname
  </label>
  <input
    className="form-control"
    id="inputSurname"
    onChange={[Function]}
    placeholder="Surname"
    required={true}
    type="text"
    value=""
  />
  <p>
    Back to 
    <a
      href="/login"
      onClick={[Function]}
    >
      <strong>
        Login page
      </strong>
    </a>
  </p>
  <button
    className="btn btn-lg btn-primary btn-block"
    onClick={[Function]}
    type="submit"
  >
    Sign up
  </button>
</form>
`;
